uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/leave-widget/leave-widget-absence-types-available-balance.component.js

```diff
@@ -37,11 +36,16 @@ define([
 
     /**
      * Implements the $onChanges controller method. It watches for changes in
-     * the component bindings.
+     * the component bindings. After bindings are ready, it loads dependencies,
+     * maps absence types with their entitlements, and then filters only absence
+     * types the contact is entitled to.
      */
     function $onChanges () {
       if (areBindingsReady()) {
-        loadDependencies();
+        loadDependencies().then(function () {
```

Comment:
i checked the sequence of promises, and there could be definitely improvements in the entire flow

```js
function $onChanges () {
  if (!areBindingsReady()) {
    return;
  }

  // @NOTE: loadDependencies() was pointless since it was just calling
  // loadEntitlements()
  loadEntitlements()
    .then(function (entitlements) {
      return _.indexBy(entitlements, 'type_id');
    })
    .then(mapAbsenceTypesWithTheirEntitlements)
    .then(filterAbsenceTypes) // @NOTE come up with a better name please
    .then(function (filteredAbsenceTypes) {
      // @NOTE: only now that the list is ready, you assign it to the scope
      vm.absenceTypes = filteredAbsenceTypes;
    })
    .then(function () {
      // @NOTE: the child is ready only when all the data is ready
      // and the event is explicitly sent from the main function
      $scope.$emit('LeaveWidget::childIsReady', childComponentName);
    });
}

function filterAbsenceTypes (absenceTypesWithEntitlements) {
  return absenceTypesWithEntitlements.filter(/* .. */)
}

function loadEntitlements () {
  return Entitlement.all({
    'contact_id': vm.contactId,
    'period_id': vm.absencePeriod.id,
    'type_id.is_active': true
  }, true);
}

/**
 * Maps each absence type with their corresponding entitlement.
 */
function mapAbsenceTypesWithTheirEntitlements (mapAbsenceTypesWithTheirEntitlements) {
  return vm.absenceTypes.map(/* .. */);
}
```